Bottom: f1b4cae5cf2967504e861f6f7637d293f698ad01
Top:    e5661e871a8d8518bb5cfcd63df265ce97d73cea
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-07-12 11:12:22 -0500

Populate buildbot with gitlab public keys


---
diff --git a/ansible-playbook/buildbot.yml b/ansible-playbook/buildbot.yml
index e14b93b9..f5bf035e 100644
--- a/ansible-playbook/buildbot.yml
+++ b/ansible-playbook/buildbot.yml
@@ -6,15 +6,20 @@
   become_user: root
   become_method: sudo
   tags:
-  - buildbot
-  roles:
-  - name: Build and start the Buildbot master container
-    role: buildbot-master
-    tags:
+    - buildbot
     - buildbot-master
-
+  tasks:
+  - name: Scan Gitlab ssh keys
+    command: ssh-keyscan {{gitlab_hostname}}
+    register: gitlab_keys
+  - name: Build and start the Buildbot master container
+    import_role: 
+      name: buildbot-master
+    vars:
+      known_hosts: "{{gitlab_keys.stdout}}"
   - name: Add buildbot user to Gitlab
-    role: gitlab-user
+    import_role: 
+      name: gitlab-user
     vars:
       new_user: buildbot
       new_user_name: Build Robot
@@ -27,11 +32,11 @@
   become_user: root
   become_method: sudo
   tags:
-  - buildbot
+    - buildbot
   roles:
   - name: Add buildbot_master pubkey to buildbot user
     role: gitlab-sshkey
     vars:
       username: buildbot
-      pubkey: "{{ hostvars.buildbot_master.pubkey }}"
+      pubkey: "{{ hostvars['buildbot-master'].pubkey }}"
       title: "buildbot_master"
diff --git a/docker/buildbot-master/buildbot/start.sh b/docker/buildbot-master/buildbot/start.sh
index 85c28248..7b5acf05 100755
--- a/docker/buildbot-master/buildbot/start.sh
+++ b/docker/buildbot-master/buildbot/start.sh
@@ -8,6 +8,12 @@ BUILDBOT_DATA=${BUILDBOT_DATA-"/buildbot"}
 
 ssh-keygen -q -b ${SSH_BITS} -t ${SSH_TYPE} -N '' -f ${SSH_FILE} 0>&- 1>/dev/null 2>/dev/null
 
+if [ -n "$BUILDBOT_KNOWN_HOSTS_FILE" ] && [ -f "$BUILDBOT_KNOWN_HOSTS_FILE" ]
+then
+    mkdir -p "${HOME}/.ssh"
+    cp -v "$BUILDBOT_KNOWN_HOSTS_FILE" "${HOME}/.ssh/known_hosts"
+fi
+
 ln -v -s -f ${HOME}/buildbot.tac ${HOME}/master.cfg ${HOME}/*.py $BUILDBOT_DATA
 
 # Run original buildbot entrypoint
diff --git a/roles/buildbot-master/defaults/main.yml b/roles/buildbot-master/defaults/main.yml
index 0ef0331c..0b19fb38 100644
--- a/roles/buildbot-master/defaults/main.yml
+++ b/roles/buildbot-master/defaults/main.yml
@@ -33,6 +33,7 @@ config_hostdir: '{{ docker_restore_config_base_dir }}/{{ container_name }}'
 # environment passed to the container
 container_env:
   BUILDBOT_DATABASE: 'sqlite:///{{ data_path }}/state.sqlite'
+  BUILDBOT_KNOWN_HOSTS_FILE: '{{secrets_path}}/{{secrets_known_hosts}}'
   BUILDBOT_SMTP_HOST: 'relay.dnsexit.com'
   BUILDBOT_SMTP_USER: 'Dean.cox'
   BUILDBOT_SMTP_PASS: 'no5156'
@@ -60,7 +61,7 @@ container_networks:
 # volumes mounted within the container
 container_volumes:
   - '{{ data_volume }}:{{ data_path }}:z'
-  - '{{ config_hostdir }}:{{ secrets_path }}'
+  - '{{ secrets_hostdir }}/{{secrets_known_hosts}}:{{ secrets_path }}/{{secrets_known_hosts}}'
 
 # mount path of the data volume
 data_path: '/buildbot'
@@ -69,7 +70,7 @@ data_path: '/buildbot'
 data_volume: '{{ docker_name_prefix }}buildbot_DV'
 
 # hostname
-hostname: buildbot-worker-ntel
+hostname: '{{ role_name }}'
 
 # build arguments of the image
 image_args:
diff --git a/roles/buildbot-master/tasks/main.yml b/roles/buildbot-master/tasks/main.yml
index b1b843ba..4cc67d20 100644
--- a/roles/buildbot-master/tasks/main.yml
+++ b/roles/buildbot-master/tasks/main.yml
@@ -10,6 +10,16 @@
   docker_volume:
     name: '{{ data_volume }}'
 
+- name: Create secrets volume
+  file:
+    path: '{{ secrets_hostdir }}'
+    state: directory
+
+- name: Populate known_hosts secret
+  copy:
+    content: '{{ known_hosts }}'
+    dest: '{{ secrets_hostdir }}/{{ secrets_known_hosts }}'
+
 - name: Checkout image repo
   git:
     repo: '{{ image_repo }}'
@@ -27,7 +37,7 @@
 - name: Start buildbot container
   docker_container:
     name: '{{ container_name }}'
-    hostname: '{{ buildbot_hostname }}'
+    hostname: '{{ hostname }}'
     image: '{{ image_name }}:{{ docker_image_tag }}'
     networks: '{{ container_networks }}'
     volumes: '{{ container_volumes }}'
@@ -43,6 +53,6 @@
 
 - name: Add buildbot_master host
   add_host:
-    name: buildbot_master
+    name: "{{ hostname }}"
     pubkey: "{{ output.stdout }}"
     ip_addr: "{{ buildbot_ip_addr }}"
